iceoryx_utils/include/iceoryx_utils/cxx/unique_typed_id.hpp

```diff
@@ -0,0 +1,127 @@
+///     std::vector<MyClass> mySecondClassVector;
+///
+///     uint64_t AddClass() {
+///         myClassVector.emplace_back();
+///         return myClassVector.back().id.getID();
+///     }
+///
+///     uint64_t AddSecondClass() {
+///         mySecondClassVector.emplace_back();
+///         // we use the uint64_t conversion
+///         return mySecondClassVector.back().id;
+///     }
+///
+///     void RemoveClass(const uint64_t id) {
+///         auto iter = std::find_if(myClassVector.begin(), myClassVector.end(), [&](MyClass & c){ return c.id == id;});
+///         if ( iter != myClassVector.end() ) myClassVector.erase(iter);
+///     }
+///
+///     uint64_t id = AddClass();
+///     RemoveClass(id);
+///
+///     // it can be that id == id2 since the id is unique per type
+///     uint64_t id2 = AddSecondClass();
+/// @endcode
+template <typename T>
+class UniqueTypedID
+{
+  public:
+    /// @brief the constructor creates an id which is greater then the
+    ///         previous created id
+    UniqueTypedID() noexcept;
+
+    /// @brief move constructs a new id and invalidates rhs by setting
+    ///         the id to INVALID_ID
+    UniqueTypedID(UniqueTypedID&& rhs) noexcept;
+
+    /// @brief move assigns rhs and invalidates rhs by setting
+    ///         the id to INVALID_ID
+    UniqueTypedID& operator=(UniqueTypedID&& rhs) noexcept;
+
+    UniqueTypedID(const UniqueTypedID& rhs) = default;
+    UniqueTypedID& operator=(const UniqueTypedID& rhs) = default;
+    ~UniqueTypedID() = default;
+
+    /// @brief if two ids are equal return true, otherwise false
+    bool operator==(const UniqueTypedID& rhs) const noexcept;
+
+    /// @brief if two ids are not equal return true, otherwise false
+    bool operator!=(const UniqueTypedID& rhs) const noexcept;
+
+    /// @brief if the id of this is less then rhs return true, otherwise false
+    bool operator<(const UniqueTypedID& rhs) const noexcept;
+
+    /// @brief if the id of this is less or equal then rhs return true, otherwise false
+    bool operator<=(const UniqueTypedID& rhs) const noexcept;
+
+    /// @brief if the id of this is greater or equal then rhs return true, otherwise false
+    bool operator>=(const UniqueTypedID& rhs) const noexcept;
+
+    /// @brief if the id of this is greater then rhs return true, otherwise false
+    bool operator>(const UniqueTypedID& rhs) const noexcept;
+
+    /// @brief converts the id into a uint64_t which contains the actual
+    ///         value of the id
+    operator uint64_t() const noexcept;
+
+    /// @brief returns the value of the id
+    uint64_t getID() const noexcept;
+
+    /// @brief constant which is used to signal an invalid id. used by the
+    ///         move operations to notify the user that their object is no
+    ///         longer valid.
+    static constexpr uint64_t INVALID_ID = std::numeric_limits<uint64_t>::max();
+
+  private:
+    static std::atomic<uint64_t> globalIDCounter; // = 0u
```

Comment:
Initialization required? If so, please use `{}`.

Reply:
It is a static and cannot be initialized in place. Therefore I have to initialize it outside of the class definition and as a comment I wrote the value to what the counter is initialized to.

Reply:
This can become potentially inconsistent. Better leave out such comments